### ISSUE TITLE
feat: build standalone app

### ISSUE THREAD

**Issue by flamby** [WSH032/pytauri#2]
Hello,

Thanks for this great project, I successfully tested the pytauri_demo app on MacOS and Ubuntu.

I wonder if it's possible to change the app name in the menu bar, as well as content of the about modal window?
- Python in menu bar
- About Python
- About Modal

<img width="795" alt="Capture d’écran 2024-12-05 à 11 11 01" src="https://github.com/user-attachments/assets/10b4fd11-8694-4798-97c4-4ab898d5803e">

Keep the good works!

**Comment by WSH032** [comment]
> I wonder if it's possible to change the app name in the menu bar, as well as the content of the about modal window?

## TLDR

I have planned to implement it within a few weeks.

## Details

The launch command `python -m pytauri_demo` essentially executes a piece of Python code, which means the actual binary executable is `/bin/python`. This is why you see the `Python` field in the menu bar and about modal. To solve this issue, you need to generate your own executable file, so you can set the `identifier`, `icon`, `about`, etc., for your executable.

For now, you can use [pyinstaller](https://github.com/pyinstaller/pyinstaller) or [Nuitka](https://github.com/Nuitka/Nuitka) to package the Python interpreter to generate your own executable. For example, pyinstaller has some configuration options that allow you to [set some executable file information](https://pyinstaller.org/en/stable/usage.html#mac-os-specific-options).

However, this is not the ideal solution. In fact, pyo3 allows us to [embed the Python interpreter in an executable](https://pyo3.rs/v0.23.3/building-and-distribution.html#embedding-python-in-rust), and tauri-cli can [generate an executable](https://tauri.app/distribute/). Combining these two can achieve this functionality.

Using pyo3 to embed Python requires the executable to link to the Python dynamic library, which means:

- Either require the user to install Python
- Or bundle and distribute a portable Python to user

I prefer the second option, and I will use [python-build-standalone](https://github.com/indygreg/python-build-standalone) as the portable Python.

I have experimented with this on Windows, and it works fine. But I lack experience with Unix, so I am not sure how to bundle Python on Unix.

Specifically, on Windows, you just need to place the executable and dynamic library in the same directory.

```tree
├── python3.dll
├── python.exe
├── pytauri-demo.exe
```

But on Unix, the executable layout is:

```tree
├── bin
│   ├── python
│   └── pytauri-demo
└── lib
    └── libpython3.so
```

I am not sure how to configure tauri-cli to package with this layout. I need to ask the tauri developers for advice. There is a related issue tauri-apps/tauri#6802.

**Comment by flamby** [comment]
@WSH032 Thanks for the clarification. 

I did not know Nuitka, nor python-build-standalone. 
I'll try packaging with pyinstaller first. 

Bundling and distribute a portable Python to user is definitely mandatory for most projects to ease UX.

Thanks!

**Comment by WSH032** [comment]
> I did not know Nuitka, nor python-build-standalone.

Don't worry, these are just implementation details for those who are interested.

> I'll try packaging with pyinstaller first.

Unless you just want to experiment, I would recommend waiting for me to implement this feature. By generating an executable with tauri-cli, we can not only configure the executable file but also enjoy other features of tauri, such as the [updater](https://tauri.app/plugin/updater/) and [bundler](https://tauri.app/distribute/#bundling).

> Bundling and distribute a portable Python to user is definitely mandatory for most projects to ease UX.

Thanks for your feedback 😉

**Comment by WSH032** [comment]
I implemented support for Windows standalone building in commit d5952b3c9d0abe1418c93fdbf4e6dcbab72760c9.

Also, please follow #3.

**Comment by WSH032** [comment]
This feature has been implemented, [here is the documentation](https://wsh032.github.io/pytauri/usage/tutorial/build-standalone/).

@flamby, would you like to give it a try? If there are no other issues, I think we can close this issue.

**Comment by flamby** [comment]
Hi @WSH032 

I've successfully followed the standalone build part of the tutorial:
<img width="796" alt="Capture d’écran 2025-01-04 à 08 04 14" src="https://github.com/user-attachments/assets/c393a6f3-597b-42a4-af6a-5c827b3115bc" />

BTW, the [example_app](https://github.com/WSH032/pytauri/tree/main/example/tauri-app) seems broken to me now.
I suggest you put it back in the README.md as it was very easy to set up when one discovers your project.

Thanks!

**Comment by WSH032** [comment]
> I've successfully followed the standalone build part of the tutorial

Great! I am glad to hear that.

> BTW, the [example_app](https://github.com/WSH032/pytauri/tree/main/example/tauri-app) seems broken to me now.

Can you provide me with some specific error messages? The `example_app` is a member of the workspace (e.g., its dependencies are specified with `workspace = true` instead of `version`), so it might not be entirely consistent with the usage method in the tutorial.

> I suggest you put it back in the README.md as it was very easy to set up when one discovers your project.

Do you mean an example like this [out-of-the-box](https://github.com/WSH032/pytauri/tree/b981a1bd07efd10a53461c2675df70afbb3bad3e?tab=readme-ov-file#install)?

> Ideally, having a `pnpm create pytauri-app` would ease a lot the learning curve for newbies.

I would really like to have something like that, but currently, I don't have the time to implement it. Pytauri currently lacks many Tauri APIs, so the priority is to complete these APIs.

So I think what we can do now is to create a feature request issue and see if anyone in the community is willing to implement this feature.

Another suboptimal method is to create a repository similar to [vite-tauri-template](https://github.com/yooneskh/vite-tauri-template) and provide some Python scripts to help with initialization.

**Comment by flamby** [comment]
> Can you provide me with some specific error messages? The `example_app` is a member of the workspace (e.g., its dependencies are specified with `workspace = true` instead of `version`), so it might not be entirely consistent with the usage method in the tutorial.

```bash
(base) ~/dev rm -rf pytauri/
(base) ~/dev git clone https://github.com/WSH032/pytauri.git
Cloning into 'pytauri'...
remote: Enumerating objects: 1337, done.
remote: Counting objects: 100% (275/275), done.
remote: Compressing objects: 100% (133/133), done.
remote: Total 1337 (delta 168), reused 200 (delta 125), pack-reused 1062 (from 2)
Receiving objects: 100% (1337/1337), 866.89 KiB | 24.08 MiB/s, done.
Resolving deltas: 100% (631/631), done.
(base) ~/dev cd pytauri/
l
(base) ~/dev/pytauri main ls
CHANGELOG.md		Cargo.toml		_typos.toml		docs			mkdocs.yml		pnpm-workspace.yaml	release.py		utils
CONTRIBUTING.md		LICENSE			commitlint.config.ts	docs_src		package.json		pyproject.toml		scripts			uv.lock
Cargo.lock		README.md		crates			example			pnpm-lock.yaml		python			tsconfig.base.json
(base) ~/dev/pytauri main pnpm install
Scope: all 3 workspace projects
Lockfile is up to date, resolution step is skipped
Packages: +175
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
Progress: resolved 175, reused 175, downloaded 0, added 175, done

devDependencies:
+ @commitlint/types 19.5.0
+ @rollup/plugin-node-resolve 15.3.1
+ @rollup/plugin-replace 6.0.2
+ @rollup/plugin-terser 0.4.4
+ @rollup/plugin-typescript 12.1.2
+ @tauri-apps/api 2.1.1
+ @types/node 22.10.2
+ eslint 9.17.0
+ eslint-config-prettier 9.1.0
+ eslint-plugin-security 3.0.1
+ prettier 3.4.2
+ pyright 1.1.391
+ rollup 4.29.1
+ tauri-plugin-pytauri-api 0.1.0-beta.0 <- crates/tauri-plugin-pytauri/guest-js
+ tslib 2.8.1
+ typescript 5.7.2
+ typescript-eslint 8.18.1

Done in 2.5s
(base) ~/dev/pytauri main pnpm -r run build
Scope: 2 of 3 workspace projects
crates/tauri-plugin-pytauri/guest-js build$ rollup -c --configPlugin typescript
│
│ src/index.ts → ./dist/index.js, ./dist/index.cjs...
│ created ./dist/index.js, ./dist/index.cjs in 536ms
│
│ src/api-iife.ts → dist/api-iife.prod.js...
│ created dist/api-iife.prod.js in 572ms
│
│ src/api-iife.ts → dist/api-iife.dev.js...
│ created dist/api-iife.dev.js in 468ms
└─ Done in 4.2s
example/tauri-app build$ tsc && vite build
│ vite v6.0.5 building for production...
│ transforming...
│ ✓ 7 modules transformed.
│ rendering chunks...
│ computing gzip size...
│ src-tauri/frontend/index.html                 8.32 kB │ gzip: 3.47 kB
│ src-tauri/frontend/assets/index-pSGieOOU.css  1.37 kB │ gzip: 0.65 kB
│ src-tauri/frontend/assets/index-QZ9CWKH-.js   1.50 kB │ gzip: 0.79 kB
│ ✓ built in 89ms
└─ Done in 2.2s
(base) ~/dev/pytauri main uv venv
Using CPython 3.11.7 interpreter at: /Library/Frameworks/Python.framework/Versions/3.11/bin/python3
Creating virtual environment at: .venv
Activate with: source .venv/bin/activate
(base) ~/dev/pytauri main source .venv/bin/activate  # bash/zsh
(pytauri) (base) ~/dev/pytauri main uv pip install setuptools setuptools-rust setuptools-scm
Resolved 5 packages in 148ms
Installed 5 packages in 38ms
 + packaging==24.2
 + semantic-version==2.10.0
 + setuptools==75.6.0
 + setuptools-rust==1.10.2
 + setuptools-scm==8.1.0
(pytauri) (base) ~/dev/pytauri main uv sync \
→     --only-group=example \
→     --no-build-isolation-package=pytauri-demo \
→     --reinstall-package=pytauri-demo
Resolved 87 packages in 11ms
   Built tauri-app @ file:///Users/angusyoung/dev/pytauri/example/tauri-app/src-tauri
   Built pytauri @ file:///Users/angusyoung/dev/pytauri/python/pytauri
   Built pytauri-plugin-notification @ file:///Users/angusyoung/dev/pytauri/python/pytauri-plugin-notification
   Built pyo3-utils @ file:///Users/angusyoung/dev/pytauri/python/pyo3-utils
Prepared 4 packages in 3.61s
Installed 13 packages in 10ms
 + annotated-types==0.7.0
 + anyio==4.7.0
 + idna==3.10
 + importlib-metadata==8.5.0
 + pydantic==2.10.4
 + pydantic-core==2.27.2
 + pyo3-utils==0.1.0b1 (from file:///Users/angusyoung/dev/pytauri/python/pyo3-utils)
 + pytauri==0.1.0b0 (from file:///Users/angusyoung/dev/pytauri/python/pytauri)
 + pytauri-plugin-notification==0.1.0b0 (from file:///Users/angusyoung/dev/pytauri/python/pytauri-plugin-notification)
 + sniffio==1.3.1
 + tauri-app==0.1.0 (from file:///Users/angusyoung/dev/pytauri/example/tauri-app/src-tauri)
 + typing-extensions==4.12.2
 + zipp==3.21.0
(pytauri) (base) ~/dev/pytauri main python -m pytauri_demo
/Users/angusyoung/dev/pytauri/.venv/bin/python: No module named pytauri_demo
```
 
> Do you mean an example like this [out-of-the-box](https://github.com/WSH032/pytauri/tree/b981a1bd07efd10a53461c2675df70afbb3bad3e?tab=readme-ov-file#install)?

Yes, exactly.
 
> I would really like to have something like that, but currently, I don't have the time to implement it. Pytauri currently lacks many Tauri APIs, so the priority is to complete these APIs.

Makes sense.
 
> So I think what we can do now is to create a feature request issue and see if anyone in the community is willing to implement this feature.
> 
> Another suboptimal method is to create a repository similar to [vite-tauri-template](https://github.com/yooneskh/vite-tauri-template) and provide some Python scripts to help with initialization.

Yes as a side project is probably a good idea.

**Comment by WSH032** [comment]
> python -m pytauri_demo

The issue lies here, in `0.1.0-beta`, the example project name has changed from `pytauri-demo` to `tauri-app`.

Here is the latest usage method:

```bash
#!/bin/bash

git clone https://github.com/WSH032/pytauri.git
cd pytauri

pnpm install
# build frontend assets
pnpm -r run build

uv venv
source .venv/bin/activate
# or powershell: .venv\Scripts\Activate.ps1

uv pip install --reinstall -e example/tauri-app/src-tauri

python -m tauri_app
```

**Comment by flamby** [comment]
> > python -m pytauri_demo
> 
> The issue lies here, in `0.1.0-beta`, the example project name has changed from `pytauri-demo` to `tauri-app`.
> 
> Here is the latest usage method:
> 
> ```shell
> #!/bin/bash
> 
> git clone https://github.com/WSH032/pytauri.git
> cd pytauri
> 
> pnpm install
> # build frontend assets
> pnpm -r run build
> 
> uv venv
> source .venv/bin/activate
> # or powershell: .venv\Scripts\Activate.ps1
> 
> uv pip install --reinstall -e example/tauri-app/src-tauri
> 
> python -m tauri_app
> ```

My bad.
```bash
uv pip install --reinstall -e example/tauri-app/src-tauri
Resolved 13 packages in 132ms
   Built pyo3-utils @ file:///Users/angusyoung/dev/pytauri/python/pyo3-utils
   Built tauri-app @ file:///Users/angusyoung/dev/pytauri/example/tauri-app/src-tauri
   Built pytauri-plugin-notification @ file:///Users/angusyoung/dev/pytauri/python/pytauri-plugin-notification
   Built pytauri @ file:///Users/angusyoung/dev/pytauri/python/pytauri
Prepared 13 packages in 2.87s
Installed 13 packages in 11ms
 + annotated-types==0.7.0
 + anyio==4.7.0
 + idna==3.10
 + importlib-metadata==8.5.0
 + pydantic==2.10.4
 + pydantic-core==2.27.2
 + pyo3-utils==0.1.0b1 (from file:///Users/angusyoung/dev/pytauri/python/pyo3-utils)
 + pytauri==0.1.0b0 (from file:///Users/angusyoung/dev/pytauri/python/pytauri)
 + pytauri-plugin-notification==0.1.0b0 (from file:///Users/angusyoung/dev/pytauri/python/pytauri-plugin-notification)
 + sniffio==1.3.1
 + tauri-app==0.1.0 (from file:///Users/angusyoung/dev/pytauri/example/tauri-app/src-tauri)
 + typing-extensions==4.12.2
 + zipp==3.21.0
(pytauri) (base) ~/dev/pytauri main python -m tauri_app
Traceback (most recent call last):
  File "<frozen runpy>", line 189, in _run_module_as_main
  File "<frozen runpy>", line 148, in _get_module_details
  File "<frozen runpy>", line 112, in _get_module_details
  File "/Users/angusyoung/dev/pytauri/example/tauri-app/src-tauri/python/tauri_app/__init__.py", line 9, in <module>
    from pytauri import (
  File "/Users/angusyoung/dev/pytauri/python/pytauri/src/pytauri/__init__.py", line 3, in <module>
    from pytauri.ffi import (
  File "/Users/angusyoung/dev/pytauri/python/pytauri/src/pytauri/ffi/__init__.py", line 8, in <module>
    from pytauri.ffi._ext_mod import EXT_MOD
  File "/Users/angusyoung/dev/pytauri/python/pytauri/src/pytauri/ffi/_ext_mod.py", line 66, in <module>
    EXT_MOD = _load_ext_mod()
              ^^^^^^^^^^^^^^^
  File "/Users/angusyoung/dev/pytauri/python/pytauri/src/pytauri/ffi/_ext_mod.py", line 33, in _load_ext_mod
    ext_mod = eps[0].load()
              ^^^^^^^^^^^^^
  File "/Users/angusyoung/dev/pytauri/.venv/lib/python3.11/site-packages/importlib_metadata/__init__.py", line 189, in load
    module = import_module(match.group('module'))
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ModuleNotFoundError: No module named 'tauri_app.ext_mod'
```

**Comment by WSH032** [comment]
can you run:

```bash
echo "$PYTAURI_STANDALONE"
echo "$VIRTUAL_ENV"
python -c "import sys; print(sys.version)"
ls example/tauri-app/src-tauri/python/tauri_app
```

**Comment by flamby** [comment]
```bash
echo "$PYTAURI_STANDALONE"
1
(pytauri) (base) ~/dev/pytauri main echo "$VIRTUAL_ENV"
/Users/angusyoung/dev/pytauri/.venv
(pytauri) (base) ~/dev/pytauri main python -c "import sys; print(sys.version)"
3.11.7 (v3.11.7:fa7a6f2303, Dec  4 2023, 15:22:56) [Clang 13.0.0 (clang-1300.0.29.30)]
(pytauri) (base) ~/dev/pytauri main ls example/tauri-app/src-tauri/python/tauri_app
__init__.py	__main__.py	__pycache__	py.typed
```

**Comment by WSH032** [comment]
> same with the PYTAURI_STANDALONE env variable disabled

need reinstall it:

```bash
export PYTAURI_STANDALONE=0
uv pip install --reinstall -e example/tauri-app/src-tauri
python -m tauri_app
```

**Comment by flamby** [comment]
Yes indeed, my bad again.
Thanks!

**Comment by WSH032** [comment]
You are welcome😁
